### PR TITLE
FindCereal.cmake

### PR DESCRIPTION
**Pull request by OXPHOS**

### CODE REVIEW

**Comment by OXPHOS** [comment]
@vigsterkr 
One way I can think of to do serialization without std library:
`SGVector.h
```
		template<class Archive>
		void save(Archive & ar) const
		{
				ar(vlen);
				for (index_t i = 0; i < vlen; ++i)
					ar(*(vector+i));
		}

		template<class Archive>
		void load(Archive & ar)
		{
				unref();
				ar(vlen);
				vector = SG_MALLOC(T, vlen);
				for (index_t i = 0; i < vlen; ++i)
				{
					ar(vector[i]);
				}
		}
```

File: src/shogun/clustering/KMeans.h [hunk 1]
```diff
@@ -28,10 +28,10 @@ class CDistanceMachine;
 enum EKMeansMethod
 {
 	/** Mini batch based training */
-    KMM_MINI_BATCH,
+	KMM_MINI_BATCH,
```
Comment (vigsterkr) [inline src/shogun/clustering/KMeans.h]: please avoid committing indent changes... when the code isn't yours

File: src/shogun/lib/SGVector.h [hunk 2]
```diff
@@ -474,7 +473,9 @@ template<class T> class SGVector : public SGReferencedData
 		 * @return matrix
 		 */
 		static void convert_to_matrix(T*& matrix, index_t nrows, index_t ncols, const T* vector, int32_t vlen, bool fortran_order);
+
```
Comment (vigsterkr) [inline src/shogun/lib/SGVector.h]: this is a NOP, plz avoid doing changes like this

**Comment by vigsterkr** [comment]
@OXPHOS yeah this is better )

**Comment by vigsterkr** [comment]
@OXPHOS in the gist you've created you had a `cmake/external/cereal.cmake`. could you add that here as well?

File: CMakeLists.txt [hunk 3]
```diff
@@ -50,6 +50,7 @@ if(CCACHE_FOUND AND ENABLE_CCACHE)
 	SET(CMAKE_CXX_FLAGS "${CCACHE_FLAGS} ${CMAKE_CXX_FLAGS}")
 endif()
 
+FIND_PACKAGE(Cereal REQUIRED)
```
Comment (vigsterkr) [inline CMakeLists.txt]: this should rather be like when looking for Eigen..... if it's available on the system use the system one, if not, then download it.

File: CMakeLists.txt [hunk 4]
```diff
@@ -606,6 +616,7 @@ FIND_PACKAGE(Eigen3 ${EIGEN_VERSION_MINIMUM})
 IF(NOT EIGEN3_FOUND)
 	include(external/Eigen3)
 	LIST(APPEND SYSTEM_INCLUDES ${EIGEN_INCLUDE_DIR})
+	MESSAGE(STATUS: ${CEREAL_INCLUDE_DIR})
```
Comment (vigsterkr) [inline CMakeLists.txt]: remove this....

File: CMakeLists.txt [hunk 5]
```diff
@@ -50,6 +50,16 @@ if(CCACHE_FOUND AND ENABLE_CCACHE)
 	SET(CMAKE_CXX_FLAGS "${CCACHE_FLAGS} ${CMAKE_CXX_FLAGS}")
 endif()
 
+# Cereal Serialization Library
+FIND_PACKAGE(Cereal)
+IF(NOT CEREAL_FOUND)
+	include(external/Cereal)
+	LIST(APPEND INCLUDES ${CEREAL_INCLUDE_DIRS})
+	MESSAGE(STATUS: ${CEREAL_INCLUDE_DIR})
```
Comment (vigsterkr) [inline CMakeLists.txt]: once you are done with debugging remove this

File: CMakeLists.txt [hunk 6]
```diff
@@ -1150,7 +1161,7 @@ IF(ENABLE_TESTING)
 		IF (NOT TRAVIS_DISABLE_UNIT_TESTS AND EXISTS ${CMAKE_SOURCE_DIR}/tests/unit)
 			add_subdirectory(${CMAKE_SOURCE_DIR}/tests/unit)
 		ENDIF()
-		
+
```
Comment (vigsterkr) [inline CMakeLists.txt]: no need for this

File: cmake/external/Cereal.cmake [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+include(ExternalProject)
+ExternalProject_Add(
+    Cereal
+    PREFIX ${CMAKE_SOURCE_DIR}/Cereal
+    DOWNLOAD_DIR ${THIRD_PARTY_DIR}/Cereal
+    URL https://github.com/USCiLab/cereal/archive/v1.1.2.tar.gz
+    URL_MD5 34d4ad174acbff005c36d4d10e48cbb9
+	)
+
+SET(CEREAL_INCLUDE_DIRS ${THIRD_PARTY_DIR}/include/cereal)
```
Comment (vigsterkr) [inline cmake/external/Cereal.cmake]: please do check the content of this directory...

File: cmake/external/Cereal.cmake [hunk 8]
```diff
@@ -0,0 +1,16 @@
+GetCompilers()
+
+include(ExternalProject)
+ExternalProject_Add(
+    Cereal
+    PREFIX ${CMAKE_SOURCE_DIR}/Cereal
```
Comment (vigsterkr) [inline cmake/external/Cereal.cmake]: CMAKE_BINARY_DIR please

File: cmake/external/Cereal.cmake (hunk 7 above)
Comment (vigsterkr) [inline cmake/external/Cereal.cmake]: so this should work
SET(CEREAL_INCLUDE_DIRS ${CMAKE_BINARY_DIR}/Cereal/src/Cereal/include)

File: cmake/external/Cereal.cmake [hunk 9]
```diff
@@ -0,0 +1,16 @@
+GetCompilers()
```
Comment (vigsterkr) [inline cmake/external/Cereal.cmake]: no need for this

File: cmake/external/Cereal.cmake [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    PREFIX ${CMAKE_SOURCE_DIR}/Cereal
+    DOWNLOAD_DIR ${THIRD_PARTY_DIR}/Cereal
+    URL https://github.com/USCiLab/cereal/archive/v1.1.2.tar.gz
+    URL_MD5 34d4ad174acbff005c36d4d10e48cbb9
+	)
+
+SET(CEREAL_INCLUDE_DIRS ${THIRD_PARTY_DIR}/include/cereal)
+LIST(APPEND SHOGUN_DEPENDS Cereal)
+
+UNSET(C_COMPILER)
```
Comment (vigsterkr) [inline cmake/external/Cereal.cmake]: you can remove this...

File: cmake/external/Cereal.cmake [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+UNSET(CXX_COMPILER)
```
Comment (vigsterkr) [inline cmake/external/Cereal.cmake]: no need for this

File: cmake/external/Cereal.cmake (hunk 7 above)
Comment (vigsterkr) [inline cmake/external/Cereal.cmake]: so this should work
SET(CEREAL_INCLUDE_DIRS ${CMAKE_BINARY_DIR}/Cereal/src/Cereal/include)

**Comment by vigsterkr** [comment]
@OXPHOS cool so now that this works, it's time to squash your commits into 1 big commit :)
you can read about how to do this here: https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/

if you having problems how to do it feel free to ping me.... once you are done we can continue on adding support for cereal serialization :)

**Comment by karlnapf** [comment]
As @vigsterkr is fine, I go ahead and merge this.

**Pull request merged**